pkg/provider/provider.go

```diff
@@ -18,14 +26,68 @@ func Provider() *schema.Provider {
 				Type:        schema.TypeString,
 				Required:    true,
 				DefaultFunc: schema.EnvDefaultFunc("SNOWFLAKE_USER", nil),
+				// TODO validation
+			},
+			"password": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				DefaultFunc: schema.EnvDefaultFunc("SNOWFLAKE_PASSWORD", nil),
+				Sensitive:   true,
+			},
+			"role": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("SNOWFLAKE_ROLE", nil),
 			},
 			"region": &schema.Schema{
 				Type:        schema.TypeString,
 				Required:    true,
 				DefaultFunc: schema.EnvDefaultFunc("SNOWFLAKE_REGION", "us-west-2"),
 			},
 		},
-		ResourcesMap:   map[string]*schema.Resource{},
+		ResourcesMap: map[string]*schema.Resource{
+			"snowflake_database": resources.Database(),
+		},
 		DataSourcesMap: map[string]*schema.Resource{},
+		ConfigureFunc:  ConfigureProvider,
+	}
+}
+
+func ConfigureProvider(s *schema.ResourceData) (interface{}, error) {
+	dsn, err := DSN(s)
+
+	log.Printf("[DEBUG] connecting to %s", dsn)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not build dsn for snowflake connection")
+	}
+
+	db, err := sql.Open("snowflake", dsn)
+	if err != nil {
+		return nil, errors.Wrap(err, "Could not open snowflake database.")
+	}
+
+	return db, nil
+}
+
+func DSN(s *schema.ResourceData) (string, error) {
+	account := s.Get("account").(string)
+	username := s.Get("username").(string)
+	password := s.Get("password").(string)
+	region := s.Get("region").(string)
+	role := s.Get("role").(string)
+
+	// us-west-2 is their default region, but if you actually specify that it won't trigger their default code
+	//  https://github.com/snowflakedb/gosnowflake/blob/52137ce8c32eaf93b0bd22fc5c7297beff339812/dsn.go#L61
```

Comment:
:cry:

Reply:
yeah that was an hour of my life I want back